### PR TITLE
[NFC] Add LookupConformancesInContextRequest

### PR DESCRIPTION
**Pull request by CodaFi**
The first commit separates taking conformance diagnostics as the start of a potential refactoring we should perform against the conformance lookup table.

The second commit specifically adds a request point for looking up all implied conformances in the table so we have a cache-point for evaluator-based dependency tracking.

This is the final request before evaluator-based tracking can proceed against the baseline we have in our test suite today.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci test

**Review by hamishknight — APPROVED** [review]

File: include/swift/AST/TypeCheckRequests.h [hunk 1]
```diff
@@ -2160,6 +2160,25 @@ class ClosureHasExplicitResultRequest
   bool isCached() const { return true; }
 };
 
+using ProtocolConformanceLookupResult = SmallVector<ProtocolConformance *, 2>;
+void simple_display(llvm::raw_ostream &out, ConformanceLookupKind kind);
+
+/// Lookup and expand all conformances in the given context.
+class LookupAllConformancesInContextRequest :
```
Comment (slavapestov) [inline include/swift/AST/TypeCheckRequests.h]: How about GetLocalConformancesRequest?

Also can you elaborate in the doc comment on what it means to "expand" them

Comment (CodaFi) [reply include/swift/AST/TypeCheckRequests.h]: It's not meant to be a general request to look into the conformance tables. This one is specifically about grabbing all of the dependency-relevant conformances so
- TypeCheckProtocol has a way to start checking conformances in the primary
- The evaluator has a cache-point.

Eventually, I would like to have this request call a real `GetLocalConformancesRequest`.

I will update the docs.

**Comment by CodaFi** [comment]
⛵️

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Review by DougGregor — APPROVED** [review]
This is a good simplification of the odd `getLocalConformances` API, thank you.

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**